flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/streaming/connectors/pulsar/PulsarConsumerSource.java

```diff
@@ -0,0 +1,203 @@
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Pulsar source (consumer) which receives messages from a topic and acknowledges messages.
+ * When checkpointing is enabled, it guarantees at least once processing semantics.
+ *
+ * <p>When checkpointing is disabled, it auto acknowledges messages based on the number of messages it has
+ * received. In this mode messages may be dropped.
+ */
+class PulsarConsumerSource<T> extends MessageAcknowledgingSourceBase<T, MessageId> implements PulsarSourceBase<T> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PulsarConsumerSource.class);
+
+	private final int messageReceiveTimeoutMs = 100;
+	private final String serviceUrl;
+	private final String topic;
+	private final String subscriptionName;
+	private final DeserializationSchema<T> deserializer;
+
+	private PulsarClient client;
+	private Consumer<byte[]> consumer;
+
+	private boolean isCheckpointingEnabled;
+
+	private final long acknowledgementBatchSize;
+	private long batchCount;
+	private long totalMessageCount;
+
+	private transient volatile boolean isRunning;
+
+	PulsarConsumerSource(PulsarSourceBuilder<T> builder) {
+		super(MessageId.class);
+		this.serviceUrl = builder.serviceUrl;
+		this.topic = builder.topic;
+		this.deserializer = builder.deserializationSchema;
+		this.subscriptionName = builder.subscriptionName;
+		this.acknowledgementBatchSize = builder.acknowledgementBatchSize;
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+
+		final RuntimeContext context = getRuntimeContext();
+		if (context instanceof StreamingRuntimeContext) {
+			isCheckpointingEnabled = ((StreamingRuntimeContext) context).isCheckpointingEnabled();
+		}
+
+		client = createClient();
+		consumer = createConsumer(client);
+
+		isRunning = true;
+	}
+
+	@Override
+	protected void acknowledgeIDs(long checkpointId, Set<MessageId> messageIds) {
+		if (consumer == null) {
+			LOG.error("null consumer unable to acknowledge messages");
+			throw new RuntimeException("null pulsar consumer unable to acknowledge messages");
+		}
+
+		if (messageIds.isEmpty()) {
+			LOG.info("no message ids to acknowledge");
+			return;
+		}
+
+		Map<String, CompletableFuture<Void>> futures = new HashMap<>(messageIds.size());
+		for (MessageId id : messageIds) {
+			futures.put(id.toString(), consumer.acknowledgeAsync(id));
+		}
+
+		futures.forEach((k, f) -> {
+			try {
+				f.get();
+			} catch (Exception e) {
+				LOG.error("failed to acknowledge messageId " + k, e);
+				throw new RuntimeException("Messages could not be acknowledged during checkpoint creation.", e);
+			}
+		});
+	}
+
+	@Override
+	public void run(SourceContext<T> context) throws Exception {
+		Message message;
+		while (isRunning) {
+			message = consumer.receive(messageReceiveTimeoutMs, TimeUnit.MILLISECONDS);
+			if (message == null) {
+				LOG.info("unexpected null message");
```

Comment:
What if there is few message in topic? This log may take too much disk usage. Just a advice